### ISSUE TITLE
Improve RTL Scroll Behavior Feature Detection

### ISSUE THREAD

**Issue by fred-wang** [KingSora/OverlayScrollbars#187]
As part of Chromium web interoperability effort, we are trying to switch to standard scroll coordinates for non-default writing modes ( see https://www.chromestatus.com/feature/5759578031521792 ). In particular, we are trying to check potential breakages by counting web pages setting scroll coordinate to positive values for scrollers that are supposed to use nonpositive values. This library is causing websites to trigger our use counters because of the operation `scrollbarDummyElement.scrollLeft(999)`. Can you please modify it to instead rely on scrollIntoView()?

Here is an example:

      function alwaysUseNonNegativeCoordinates() {
          // Append a RTL 1px square scroller containing two 2px squares on the
          // same line, reveal the hidden one and check the sign of scrollLeft.
          document.body.insertAdjacentHTML("beforeend", "<div style='direction: rtl; position: absolute; overflow: scroll; width: 1px; height: 1px;'><div style='width: 2px'><div style='display: inline-block; width: 1px; height: 1px;'></div><div style='display: inline-block; width: 1px; height: 1px;'></div></div></div>");
          var scroller =  document.body.lastElementChild;
          scroller.firstElementChild.firstElementChild.scrollIntoView();
          var result = scroller.scrollLeft > 0;
          document.body.removeChild(scroller);
          return result;
      }

**Comment by KingSora** [comment]
Good day!

Thats not possible, I'm afraid. This number is set intentional and I'm using it to detect the `rtl` scroll behavior of the browser. I did this, because chrome, firefox and IE are fundamentally different in terms of `rtl` scrolling. 

**Firefox behavior:**
If you have a 500px wide `rtl` div, and you scroll entirely to the end, you end up with `scrollLeft = -500`. (the origin is 0, and the max value is -500)
**Chrome behavior:**
If you have a 500px wide `rtl` div, and you scroll entirely to the end, you end up with `scrollLeft = 0`. (the origin is 500 and the max value is 0)
**IE behavior:**
If you have a 500px wide `rtl` div, and you scroll entirely to the end, you end up with `scrollLeft = 500`. (the origin is 0 and the max value is 500)

With `scrollIntoView` the detection would be wrong, as it always uses the _correct_ coordinates instead of always the _same_ coordinates.

**Comment by fred-wang** [comment]
But with `scrollIntoView` you can scroll to any position, without specifying a coordinate just by placing elements where you want. The `alwaysUseNonNegativeCoordinates()` function I suggested allows to detect the sign. I haven't tested IE but I think you can distinguish it from Chrome by checking that the original value is == 0?

**Comment by fred-wang** [comment]
So basically for your `i` you just need to read the initial position and for you `n` use something like `alwaysUseNonNegativeCoordinates()`.

**Comment by cathiechen** [comment]
```
  function alwaysUseNonNegativeCoordinates() {
      // Append a RTL 1px square scroller containing two 2px squares on the
      // same line, reveal the hidden one and check the sign of scrollLeft.
      document.body.insertAdjacentHTML("beforeend", "<div style='direction: rtl; position: absolute; overflow: scroll; width: 1px; height: 1px;'><div style='width: 500px'><div style='display: inline-block; width: 1px; height: 1px; float: left;'></div><div style='display: inline-block; width: 1px; height: 1px; float: right;'></div></div></div>");

      var scroller =  document.body.lastElementChild;
      scroller.firstElementChild.firstElementChild.scrollIntoView();
      var result = "left position: " + scroller.scrollLeft;
      scroller.firstElementChild.lastElementChild.scrollIntoView();
      result = result + "; right position: " + scroller.scrollLeft;
      document.body.removeChild(scroller);
      return result;
  }
```
How about use float to indicate the left and right position? Then you can distingish all the three type of coordinators.

**Comment by fred-wang** [comment]
I pasted the wrong content in my initial script, sorry. Here is a new version, including the additional check:

    function scroll_coordinates_behavior() {
        // Append a RTL 1px square scroller containing two 2px squares on the
        // same line, reveal the hidden one and check the sign of scrollLeft.
        document.body.insertAdjacentHTML("beforeend", "<div style='direction: rtl; position: absolute; overflow: hidden; width: 1px; height: 1px;'><div style='width: 2px'><div style='display: inline-block; width: 1px; height: 1px;'></div><div style='display: inline-block; width: 1px; height: 1px;'></div></div></div>");
        var scroller =  document.body.lastElementChild;
        var initial_coordinate = scroller.scrollLeft;
        scroller.firstElementChild.lastElementChild.scrollIntoView();
        var final_coordinate = scroller.scrollLeft;
        // Per the CSSOM specification, the standard behavior is:
        // - decreasing coordinates when scrolling leftward.
        // - nonpositive coordinates in RTL mode.
        var result = {
            "decreasing": final_coordinate < initial_coordinate,
            "nonpositive": final_coordinate < 0
        };
        document.body.removeChild(scroller);
        return result;
    }

**Comment by KingSora** [comment]
@cathiechen @fred-wang if you want to provide a complete solution, then please adapt this fiddle: https://jsfiddle.net/epbt7wuL/

The first call of the `detectBehavior` is the current behavior. If I change it with `scrollIntoView` the result isn't the same anymore (I provided a reason why this is in my first answer).

The criteria that your solution is working are the following: 
The object your solution is returning / logging must be the same as the object the `scrollLeft` solution is returning / logging. This has to be the case in FF, Chrome and IE (the object changes in all browsers)

**Comment by fred-wang** [comment]
@KingSora Have you tried https://github.com/KingSora/OverlayScrollbars/issues/187#issuecomment-586297743 ? It uses scrollIntoView, has initial scrollLeft 0 and I tested it in FF, Chrome and IE. Anyway, I'll give a try to your jsfiddle or send a PR later.

**Comment by fred-wang** [comment]
@KingSora Just to be clear, you obviously need more descendants to be able to use `scrollIntoView()`. I hope this is ok for you.

**Comment by KingSora** [comment]
> 
> 
> @KingSora Have you tried [#187 (comment)](https://github.com/KingSora/OverlayScrollbars/issues/187#issuecomment-586297743) ? It uses scrollIntoView, has initial scrollLeft 0 and I tested it in FF, Chrome and IE. Anyway, I'll give a try to your jsfiddle or send a PR later.

You code in this comment doesn't have `scrollLeft = 0` as initial value, it uses the default initial value of the browser.

By the way, would it help for your detection if I made the element invisible (opacity: 0 or visibility: hidden) and you would implement this in your detection as a exception?

**Comment by fred-wang** [comment]
> You code in this comment doesn't have `scrollLeft = 0` as initial value, it uses the default initial value of the browser.

Right sorry I was confused why you said you want scrollLeft = 0 but not the standard behavior (the CSSOM spec does say initial value is 0). The detection can do a useless scrolLeft = 0 at the beginning, but I don't get why you don't want to keep the default initial value. Can you please explain?

> By the way, would it help for your detection if I made the element invisible (opacity: 0 or visibility: hidden) and you would implement this in your detection as a exception?

I think opacity or visibility is not important for this detection.

**Comment by KingSora** [comment]
Because the initial value of `0` applies only to `ltr` elements. `rtl` elements can have the maximum value as their initial value (like in chrome at the moment): https://jsfiddle.net/4nt13aqs/

Also Firefox sometimes caches scroll values of elements and applies them after you reload your page. This is basically a nice feature for users but bad for this detection. Maybe there are other browser behaviors which alter the initial value, I dont know. Thats why I want to be safe and decide the initial value of `0` to prevent unexpected behavior.

**Comment by fred-wang** [comment]
@KingSora OK, I guess there is a misunderstanding. The feature detection is exactly to detect inconsistent behavior by actually forcing them. That's the case of your current `rtlScrollBehavior()`. And of course the script in https://github.com/KingSora/OverlayScrollbars/issues/187#issuecomment-586297743 relies on inconsistent initial values and on inconsistent sign.

Anyway, I think it's still possible to adapt your code using `scrollIntoView()`, again the only requirement is that you are happy to change the DOM structure (so you can actually call `scrollIntoView()` on an element to reveal) and maybe style (for more tweaks).

**Comment by KingSora** [comment]
@fred-wang It's possible to change the DOM, although I would be happy if the code in the end wouldn't be much longer than it is now.

**Comment by fred-wang** [comment]
For the record, I edited https://github.com/KingSora/OverlayScrollbars/issues/187#issuecomment-586297743 to use Cathie's suggestion (add an initial `scrollIntoView()` call to force the position). Still not sure whether that actually addresses the concern.

**Comment by fred-wang** [comment]
@KingSora I opened https://github.com/KingSora/OverlayScrollbars/pull/188 ; please let me know what you think about it.

**Comment by KingSora** [comment]
As discussed in #188 this problem will be fixed with the solution described in this fiddle: https://jsfiddle.net/o8pscgex/ with the `detectBehavior2` function. 

For users which use a OverlayScrollbars version which isn't fixed yet, you could apply a exception, that you ignore all elements with the id `os-dummy-scrollbar-size` (thats the ID which is used for the container to which the `scrollLeft` value gets applied. If you wan't you can further specify this exception by also adding a constraint that this element has exactly one child `div` element. In case you wan't also add some CSS checks or something you can check out the styling [here](https://github.com/KingSora/OverlayScrollbars/blob/master/css/OverlayScrollbars.css#L85).

It may be worth mentioning that `Simplebar` (another popular scrollbar library) which is doing this feature detection [in the same way as OverlayScrollbars](https://github.com/Grsmto/simplebar/blob/master/packages/simplebar/src/simplebar.js#L90), as the author used my approach as inspiration.

I'll close this issue as soon as the fix is released.

**Comment by KingSora** [comment]
This is now implemented as of  [version 1.11.0](https://github.com/KingSora/OverlayScrollbars/releases/tag/v1.11.0). :)

**Comment by fred-wang** [comment]
@KingSora Great! Thanks again!
